### PR TITLE
Fix Module backward hooks for all Tensor inputs/outputs

### LINKED ISSUE #598: Problem with backward hook function
Hi,

there is something strange in the `backward` step (or maybe something I don't understand). If I define a Module that takes 3 inputs, the `grad_input` has to be of size 3, right ? But this is not the case here (from the backward_hook point of view):

```import torch
from torch.autograd import Variable
import torch.nn as nn
import torch.nn.functional as F

def bh(m,go,gi):
    print("Grad Input")
    print(go)
    print("Grad Output")
    print(gi)

class M(nn.Module):
    def __init__(self):
        super(M,self).__init__()
        self.register_backward_hook(bh)


    def forward(self,x,y,z):
        return (x+y+z)

x=Variable(torch.randn(1,5),requires_grad=True)
y=Variable(torch.randn(1,5),requires_grad=True)
z=Variable(torch.randn(1,5),requires_grad=True)

criterion=nn.MSELoss()
mod=M()
out=mod(x,y,z)
loss=criterion(out,Variable(torch.randn(1,5)))
loss.backward()```

In that case, when I print grad_input throught the hook function, it is just composed of two elements... Could you tell me where am I wrong ? But `x.grad, y.grad and z.grad` seem correctly computed

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

### PR DESCRIPTION
**Pull request by albanD**
Fixes #598

This PR is deprecating the current `Module.register_backward_hook()` function officially. Trying to warn only when it is not working as expected and run otherwise.
The next step will be to replace the `Module.register_backward_hook()` function by the newly added `register_full_backward_hook` to only keep the correct behavior.

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1313,14 +1303,8 @@ def requires_grad_(self: T, requires_grad: bool = True) -> T:
...
 
-    def zero_grad(self, set_to_none: bool = False) -> None:
-        r"""Sets gradients of all model parameters to zero. See similar function
-        under :class:`torch.optim.Optimizer` for more context.
-
-        Arguments:
-            set_to_none (bool): instead of setting to zero, set the grads to None.
-                See :meth:`torch.optim.Optimizer.zero_grad` for details.
-        """
+    def zero_grad(self) -> None:
```
Comment (albanD) [inline torch/nn/modules/module.py:1306]: Bad copy paste. These changes should not be included here. Will revert.

**Comment by mruberry** [comment]
Lint and doc errors are real. Docs error snippet:

```
Oct 12 20:36:33 /opt/conda/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py:138: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
Oct 12 20:36:33   warnings.warn("torch.distributed.reduce_op is deprecated, please use "
Oct 12 20:36:55 /opt/conda/lib/python3.6/site-packages/torch/nn/modules/flatten.py:docstring of torch.nn.Unflatten.register_backward_hook:15: WARNING: Unknown directive type "warn".
Oct 12 20:36:55 
Oct 12 20:36:55 .. warn::
Oct 12 20:36:55     Modifying inputs or outputs inplace is not allowed when using backward hooks and
Oct 12 20:36:55     will raise an error.
Oct 12 20:36:55 /opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py:docstring of torch.nn.Module.register_backward_hook:15: WARNING: Unknown directive type "warn".
Oct 12 20:36:55 
Oct 12 20:36:55 .. warn::
Oct 12 20:36:55     Modifying inputs or outputs inplace is not allowed when using backward hooks and
Oct 12 20:36:55     will raise an error.
Oct 12 20:36:55 /opt/conda/lib/python3.6/site-packages/torch/nn/modules/flatten.py:docstring of torch.nn.Flatten.register_backward_hook:15: WARNING: Unknown directive type "warn".
Oct 12 20:36:55 
Oct 12 20:36:55 .. warn::
Oct 12 20:36:55     Modifying inputs or outputs inplace is not allowed when using backward hooks and
Oct 12 20:36:55     will raise an error.
Oct 12 20:36:55 /opt/conda/lib/python3.6/site-packages/torch/jit/_script.py:docstring of torch.jit.ScriptModule.register_backward_hook:15: WARNING: Unknown directive type "warn".
Oct 12 20:36:55 
Oct 12 20:36:55 .. warn::
Oct 12 20:36:55     Modifying inputs or outputs inplace is not allowed when using backward hooks and
Oct 12 20:36:55     will raise an error.
```

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -616,26 +616,21 @@ def register_backward_hook(
...
-        module has multiple inputs or outputs. The hook should not modify its
-        arguments, but it can optionally return a new gradient with respect to
-        input that will be used in place of :attr:`grad_input` in subsequent
-        computations. :attr:`grad_input` will only correspond to the inputs given
-        as positional arguments.
+        The :attr:`grad_input` and :attr:`grad_output` are tuples. The hook should
+        not modify its arguments, but it can optionally return a new gradient with
+        respect to the input that will be used in place of :attr:`grad_input` in
+        subsequent computations. :attr:`grad_input` will only correspond to the inputs given
+        as positional arguments. Entries in :attr:`grad_input` and :attr:`grad_output` will
```
Comment (mruberry) [inline torch/nn/modules/module.py:628]: May want to extend this sentence with an extra clause clarifying that kwarg-only arguments will not appear in the hook.

File: torch/nn/modules/_functions.py [hunk 3]
```diff
@@ -196,3 +196,13 @@ def backward(ctx, grad_output):
                 accum_ratio.add_(paddded_ratio[c], alpha=-1)
 
         return grad_input, None, None, None, None
+
+class BackwardHookFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, *args):
```
Comment (mruberry) [inline torch/nn/modules/_functions.py:202]: Why not *args, **kwargs?

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -616,26 +616,21 @@ def register_backward_hook(
...
-            hook(module, grad_input, grad_output) -> Tensor or None
+            hook(module, grad_input, grad_output) -> tuple(Tensor) or None
 
-        The :attr:`grad_input` and :attr:`grad_output` may be tuples if the
-        module has multiple inputs or outputs. The hook should not modify its
-        arguments, but it can optionally return a new gradient with respect to
-        input that will be used in place of :attr:`grad_input` in subsequent
-        computations. :attr:`grad_input` will only correspond to the inputs given
-        as positional arguments.
+        The :attr:`grad_input` and :attr:`grad_output` are tuples. The hook should
```
Comment (mruberry) [inline torch/nn/modules/module.py:624]: Can this elaborate more on what, precisely, grad_input and grad_output are?

File: torch/utils/hooks.py [hunk 5]
```diff
@@ -58,3 +61,110 @@ def warn_if_has_hooks(tensor):
                               "serialized.  If this is expected, you can "
                               "decorate the function with @torch.utils.hooks.unserializable_hook "
                               "to suppress this warning".format(repr(hook)))
+
+class BackwardHook(object):
+    def __init__(self, module, user_hooks):
```
Comment (mruberry) [inline torch/utils/hooks.py:66]: Docstring?

File: torch/utils/hooks.py [hunk 6]
```diff
@@ -58,3 +61,110 @@ def warn_if_has_hooks(tensor):
...
+        for idx in indices:
+            res.append(values[idx])
+
+        return tuple(res)
+
+    def _set_user_hook(self, grad_fn, user_hook):
+        @functools.wraps(user_hook)
+        def hook(grad_input, _):
+            if self.grad_outputs is None:
+                raise RuntimeError("Module backward hook for grad input is called "
```
Comment (mruberry) [inline torch/utils/hooks.py:94]: I'm having a hard time with this error message. Could it be more direct in explaining what the issue is and how to resolve it?

File: torch/nn/modules/_functions.py (hunk 3 above)
Comment (albanD) [reply torch/nn/modules/_functions.py:202]: kwargs are not allowed for custom Functions.

File: torch/nn/modules/module.py (hunk 2 above)
Comment (albanD) [reply torch/nn/modules/module.py:628]: This is true for all the nn.Module hooks actually. I will update.

File: torch/utils/hooks.py (hunk 6 above)
Comment (albanD) [reply torch/utils/hooks.py:94]: The thing is that I can't think of a simple case where this would happen.
The user would need to have an input that requires grad + and output that does not (or at least is not used to compute the loss) + they extract a Tensor that requires grad based on the input and use that to compute the loss.
I am not sure how to capture this easily though.

Comment (mruberry) [reply torch/utils/hooks.py:94]: Maybe you could just say something like: "Missing grad outputs. (suggested remedy goes here)." Or, since it's rare that this will occur, maybe it doesn't need to be improved.

Comment (albanD) [reply torch/utils/hooks.py:94]: Is the following understandable: "This happens because the gradient in your nn.Module flow to the input without passing by the output. Make sure that the output depends on the input and that the loss is computed based on the output."

Comment (mruberry) [reply torch/utils/hooks.py:94]: > Is the following understandable: "This happens because the gradient in your nn.Module flow to the input without passing by the output. Make sure that the output depends on the input and that the loss is computed based on the output."

Makes a lot of sense. Two minor corrections: 

"This happens because the gradient in your nn.Module flows to the input without passing through the output. Make sure that the output depends on the input and that the loss is computed based on the output."

**Comment by albanD** [comment]
@mruberry All the docs have been updated and the CI fixed. Can you have another look?

File: torch/utils/hooks.py [hunk 7]
```diff
@@ -58,3 +61,122 @@ def warn_if_has_hooks(tensor):
                               "serialized.  If this is expected, you can "
                               "decorate the function with @torch.utils.hooks.unserializable_hook "
                               "to suppress this warning".format(repr(hook)))
+
+class BackwardHook(object):
+    """
+    A wrapper class to implement nn.Module backward hooks.
+    It handles:
+      - Ignoring non-Tensor inputs and replacing the, by None before calling the user hook
```
Comment (mruberry) [inline torch/utils/hooks.py:69]: Missing something after "replace the"

**Review by mruberry — APPROVED** [review]
Made one small comment about a docstring, otherwise I think this looks good.

File: torch/utils/hooks.py (hunk 7 above)
Comment (albanD) [reply torch/utils/hooks.py:69]: Ho qwerty keyboard where `m` is `,` updated.

**Review by mruberry — APPROVED** [review]
Stamped! Looks much clearer.
